src/pkcs11/framework-pkcs15.c

```diff
@@ -3680,6 +3680,58 @@ pkcs15_prkey_get_attribute(struct sc_pkcs11_session *session,
 }
 
 
+
+static CK_RV
+pkcs15_prkey_check_pss_param(CK_MECHANISM_PTR pMechanism, CK_ULONG hlen)
+{
+	CK_RSA_PKCS_PSS_PARAMS *pss_param;
+
+	if (pMechanism->pParameter == NULL)
+		return CKR_OK;				// Support applications that don't provide CK_RSA_PKCS_PSS_PARAMS
+
+	if (pMechanism->ulParameterLen != sizeof(CK_RSA_PKCS_PSS_PARAMS))
+		return CKR_MECHANISM_PARAM_INVALID;
+
+	pss_param = (CK_RSA_PKCS_PSS_PARAMS *)pMechanism->pParameter;
+
+	// Hash parameter must match mechanisms or length of data supplied for CKM_RSA_PKCS_PSS
+	switch(pss_param->hashAlg) {
+	case CKM_SHA_1:
+		if (hlen != 20)
+			return CKR_MECHANISM_PARAM_INVALID;
+		break;
+	case CKM_SHA256:
+		if (hlen != 32)
+			return CKR_MECHANISM_PARAM_INVALID;
+		break;
+	default:
+		return CKR_MECHANISM_PARAM_INVALID;
+	}
+
+	// SmartCards typically only support MGFs based on the same hash as the
+	// message digest
```

Comment:
This limitation is not on PKCS#11 level so the errors might be confusing. Even though, this might be a good idea to implement now opensc-wise (since the sc-hsm is the only supporting PSS), future drivers might struggle with it so I would propose to do this limitation in the sc-hsm code, if you know your cards do not support these cases.

Reply:
Unfortunately this would require substantial changes in the lower layer to pass PSS parameter down into the card driver. It is something we could do in the long run, however I don't expect cards to support arbitrary salt length and MGF anytime soon. PKCS#11 allows the flexibility, but most implementation use saltlen = hashlen and MGF hash = message hash as all other cases would need to be explicitly agreed between sender and receiver.

Reply:
Thank you for clarification. Yes, it makes sense to support at least something for now.